cfd_protocol/src/interval.rs

```diff
@@ -0,0 +1,124 @@
+use anyhow::{bail, Result};
+use bit_vec::BitVec;
+use std::fmt::Display;
+use std::ops::RangeInclusive;
+
+mod digit_decomposition;
+
+const BASE: usize = 2;
+
+/// Maximum number of binary digits for BTC price in whole USD.
+const MAX_DIGITS: usize = 20;
+
+const MAX_PRICE_DEC: u64 = (BASE as u64).pow(MAX_DIGITS as u32);
+
+#[derive(Debug)]
+pub struct Interval(RangeInclusive<u64>);
+
+impl Interval {
+    pub fn new(start: u64, end: u64) -> Result<Self> {
+        if start > MAX_PRICE_DEC || end > MAX_PRICE_DEC {
+            bail!("price over maximum")
+        }
+
+        if start > end {
+            bail!("invalid interval: start > end")
+        }
+
+        Ok(Self(start..=end))
+    }
+
+    pub fn as_digits(&self) -> Vec<Digits> {
+        digit_decomposition::group_by_ignoring_digits(
+            *self.0.start() as usize,
+            *self.0.end() as usize,
+            BASE,
+            MAX_DIGITS,
+        )
+        .iter()
+        .map(|digits| {
+            let digits = digits.iter().map(|n| *n != 0).collect::<BitVec>();
+            Digits(digits)
+        })
+        .collect()
+    }
+}
```

Comment:
It looks like the public API of this module is essentially: `(start, end) -> Vec<Digits>`

You could expose that as a free function and make `Interval` private (or even inline it if deemed unnecessary).